staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go

```diff
@@ -35,72 +36,121 @@ func init() {
 type TokenGetter interface {
-	Token() (string, error)
+	Token() (*openstackToken, error)
 }
 
-type tokenGetter struct{}
+type tokenGetter struct {
+	options gophercloud.AuthOptions
+}
 
-// Token creates a token by authenticate with keystone.
-func (*tokenGetter) Token() (string, error) {
-	options, err := openstack.AuthOptionsFromEnv()
+// Token gets a token from the OpenStack Identity service
+func (tg *tokenGetter) Token() (*openstackToken, error) {
+	token, err := tokenFromAuth(tg.options)
 	if err != nil {
-		return "", fmt.Errorf("failed to read openstack env vars: %s", err)
+		return &openstackToken{}, fmt.Errorf("error getting openstack token: %s", err)
 	}
-	client, err := openstack.AuthenticatedClient(options)
-	if err != nil {
-		return "", fmt.Errorf("authentication failed: %s", err)
-	}
-	return client.TokenID, nil
+	return token, nil
 }
 
-// cachedGetter caches a token until it gets expired, after the expiration, it will
-// generate another token and cache it.
 type cachedGetter struct {
-	mutex       sync.Mutex
-	tokenGetter TokenGetter
+	lk        sync.Mutex
+	tg        TokenGetter
+	persister restclient.AuthProviderConfigPersister
+	cache     map[string]string
+}
 
-	token string
-	born  time.Time
-	ttl   time.Duration
+func newCachedGetter(tg TokenGetter, persister restclient.AuthProviderConfigPersister, cache map[string]string) *cachedGetter {
+	if cache == nil {
+		cache = make(map[string]string)
+	}
+	return &cachedGetter{
+		tg:        tg,
+		persister: persister,
+		cache:     cache,
+	}
 }
 
 // Token returns the current available token, create a new one if expired.
-func (c *cachedGetter) Token() (string, error) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-
-	var err error
-	// no token or exceeds the TTL
-	if c.token == "" || time.Now().Sub(c.born) > c.ttl {
-		c.token, err = c.tokenGetter.Token()
-		if err != nil {
-			return "", fmt.Errorf("failed to get token: %s", err)
+func (g *cachedGetter) Token() (*openstackToken, error) {
+	token := g.cachedToken()
+	if token.Valid() {
+		return token, nil
+	}
+	token, err := g.tg.Token()
+	if err != nil {
+		return nil, err
+	}
+	cache := g.update(token)
+	if g.persister != nil {
+		if err := g.persister.Persist(cache); err != nil {
+			glog.V(4).Infof("Failed to persist token: %v", err)
 		}
-		c.born = time.Now()
 	}
-	return c.token, nil
+	return token, nil
+}
+
+func (g *cachedGetter) cachedToken() *openstackToken {
+	g.lk.Lock()
+	defer g.lk.Unlock()
+	id, _ := g.cache["token-id"]
+	expiresAt, _ := g.cache["expires-at"]
+	var expiresAtTime time.Time
+	if parsedTime, err := time.Parse(time.RFC3339Nano, expiresAt); err == nil {
+		expiresAtTime = parsedTime
+	}
+	return &openstackToken{
+		ID:        id,
+		ExpiresAt: expiresAtTime,
+	}
+}
+
+func (g *cachedGetter) update(token *openstackToken) map[string]string {
+	g.lk.Lock()
+	defer g.lk.Unlock()
+	ret := map[string]string{}
+	for k, v := range g.cache {
+		ret[k] = v
+	}
```

Comment:
Is there a need to copy the cache into ret? Maybe we just can update and return g.cache directly?  Tbh I don't know why they did that for gcp https://github.com/kubernetes/client-go/blob/master/plugin/pkg/client/auth/gcp/gcp.go#L242-L244

Reply:
From what I understand, as soon as `update` returns, another goroutine can update `g.cache`. If we return `g.cache` directly, we may end up persisting the other goroutine's values, or garbage (e.g. our expires-at, but the other goroutine's token-id).

For more context, please see [the commit that introduced this](https://github.com/kubernetes/kubernetes/commit/5a6cd558c23657977ea4cb9103aa51d068c3fe57) to the gcp auth provider.